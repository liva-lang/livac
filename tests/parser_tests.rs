use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper para casos correctos del parser
fn test_parser_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();
    
    // Convertir AST a JSON para snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("ok_{}.ast", test_name), json);
}

/// Test helper para casos de error del parser
fn test_parser_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let result = parse(tokens, &source);
    assert!(result.is_err(), "Expected parser error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_functions_oneliner() {
    test_parser_ok("functions_oneliner");
}

#[test]
fn test_functions_block() {
    test_parser_ok("functions_block");
}

#[test]
fn test_classes() {
    test_parser_ok("classes");
}

#[test]
fn test_control_flow() {
    test_parser_ok("control_flow");
}

#[test]
fn test_expressions() {
    test_parser_ok("expressions");
}

#[test]
fn test_imports() {
    test_parser_ok("imports");
}

#[test]
fn test_unclosed_paren() {
    test_parser_err("unclosed_paren");
}

#[test]
fn test_unclosed_brace() {
    test_parser_err("unclosed_brace");
}

#[test]
fn test_case_without_switch() {
    test_parser_err("case_without_switch");
}

#[test]
fn test_return_outside_function() {
    test_parser_err("return_outside_function");
}

#[test]
fn test_duplicate_default() {
    test_parser_err("duplicate_default");
}
