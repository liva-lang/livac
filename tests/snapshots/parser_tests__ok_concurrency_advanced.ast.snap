---
source: tests/parser_tests.rs
expression: json
---
{
  "items": [
    {
      "UseRust": {
        "crate_name": "tokio",
        "alias": null
      }
    },
    {
      "UseRust": {
        "crate_name": "std::thread",
        "alias": null
      }
    },
    {
      "Function": {
        "name": "fetchMultipleUsers",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "user1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "fetchUser"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 1
                        }
                      }
                    ],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "user2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "fetchUser"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 2
                        }
                      }
                    ],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "user3",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "fetchUser"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 3
                        }
                      }
                    ],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Return": {
                "expr": {
                  "ArrayLiteral": [
                    {
                      "Identifier": "user1"
                    },
                    {
                      "Identifier": "user2"
                    },
                    {
                      "Identifier": "user3"
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "processData",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "data",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "fetchData"
                    },
                    "args": [],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "processed",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "processIntensive"
                    },
                    "args": [
                      {
                        "Identifier": "data"
                      }
                    ],
                    "exec_policy": "par",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "logProcessStart"
                    },
                    "args": [],
                    "exec_policy": "fire-async",
                    "type_args": []
                  }
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "task_handle",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "backgroundWork"
                    },
                    "args": [],
                    "exec_policy": "task-par",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "final",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Unary": {
                    "op": "Await",
                    "operand": {
                      "Identifier": "task_handle"
                    }
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Return": {
                "expr": {
                  "ArrayLiteral": [
                    {
                      "Identifier": "data"
                    },
                    {
                      "Identifier": "processed"
                    },
                    {
                      "Identifier": "final"
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "mixedConcurrencyDemo",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "a1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "task1"
                    },
                    "args": [],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "a2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "task2"
                    },
                    "args": [],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "a3",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "task3"
                    },
                    "args": [],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "p1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "cpuTask"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 1
                        }
                      }
                    ],
                    "exec_policy": "par",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "p2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "cpuTask"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 2
                        }
                      }
                    ],
                    "exec_policy": "par",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Binary": {
                    "op": "Add",
                    "left": {
                      "Binary": {
                        "op": "Add",
                        "left": {
                          "Identifier": "a1"
                        },
                        "right": {
                          "Identifier": "a2"
                        }
                      }
                    },
                    "right": {
                      "Identifier": "a3"
                    }
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Binary": {
                    "op": "Add",
                    "left": {
                      "Identifier": "p1"
                    },
                    "right": {
                      "Identifier": "p2"
                    }
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "logResults"
                    },
                    "args": [
                      {
                        "Identifier": "result1"
                      },
                      {
                        "Identifier": "result2"
                      }
                    ],
                    "exec_policy": "fire-async",
                    "type_args": []
                  }
                }
              }
            },
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "cleanup"
                    },
                    "args": [],
                    "exec_policy": "fire-par",
                    "type_args": []
                  }
                }
              }
            },
            {
              "Return": {
                "expr": {
                  "Binary": {
                    "op": "Mul",
                    "left": {
                      "Identifier": "result1"
                    },
                    "right": {
                      "Identifier": "result2"
                    }
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "lazyConcurrencyDemo",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "future1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "expensiveAsyncOperation"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 1
                        }
                      }
                    ],
                    "exec_policy": "async",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "future2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "expensiveParallelOperation"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 2
                        }
                      }
                    ],
                    "exec_policy": "par",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "intermediate",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "doSomeWork"
                    },
                    "args": [],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result1",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Identifier": "future1"
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result2",
                    "type_ref": null
                  }
                ],
                "init": {
                  "Identifier": "future2"
                },
                "is_fallible": false
              }
            },
            {
              "Return": {
                "expr": {
                  "Binary": {
                    "op": "Add",
                    "left": {
                      "Binary": {
                        "op": "Add",
                        "left": {
                          "Identifier": "result1"
                        },
                        "right": {
                          "Identifier": "result2"
                        }
                      }
                    },
                    "right": {
                      "Identifier": "intermediate"
                    }
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "main",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "print"
                    },
                    "args": [
                      {
                        "Literal": {
                          "String": "Advanced concurrency features parsed successfully"
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    }
  ]
}
